pipeline/measuring.py

```diff
@@ -139,166 +141,184 @@ def run(self, *args, **kwargs):
-                m.area_psf = output['psf_area']
-                m.flux_apertures = output['fluxes']
-                m.flux_apertures_err = [np.sqrt(output['variance'] * a) for a in output['areas']]  # TODO: add source noise??
-                m.aper_radii = output['radii']
-                m.area_apertures = output['areas']
-                m.background = output['background']
-                m.background_err = np.sqrt(output['variance'])
-                m.offset_x = output['offset_x']
-                m.offset_y = output['offset_y']
-                m.width = (output['major'] + output['minor']) / 2
-                m.elongation = output['elongation']
-                m.position_angle = output['angle']
-
-                if self.pars.chosen_aperture == 'auto':
-                    raise NotImplementedError('Automatic aperture selection is not yet implemented.')
-                if self.pars.chosen_aperture == 'psf':
-                    ap_index = -1
-                elif isinstance(self.pars.chosen_aperture, int):
-                    ap_index = self.pars.chosen_aperture
-                else:
-                    raise ValueError(
-                        f'Invalid value "{self.pars.chosen_aperture}" for chosen_aperture in the measuring parameters.'
+        try:  # first make sure we get back a datastore, even an empty one
+            # most likely to get a Cutouts object or list of Cutouts
+            if isinstance(args[0], Cutouts):
+                new_args = [args[0]]  # make it a list if we got a single Cutouts object for some reason
+                new_args += list(args[1:])
+                args = tuple(new_args)
+
+            if isinstance(args[0], list) and all([isinstance(c, Cutouts) for c in args[0]]):
+                args, kwargs, session = parse_session(*args, **kwargs)
+                ds = DataStore()
+                ds.cutouts = args[0]
+                ds.detections = ds.cutouts[0].sources
+                ds.sub_image = ds.detections.image
+                ds.image = ds.sub_image.new_image
+            else:
+                ds, session = DataStore.from_args(*args, **kwargs)
+        except Exception as e:
+            return DataStore.catch_failure_to_parse(e, *args)
+
+        try:
+            t_start = time.perf_counter()
+            if os.getenv('SEECHANGE_TRACEMALLOC') == '1':
+                import tracemalloc
+                tracemalloc.reset_peak()  # start accounting for the peak memory usage from here
+
+            self.pars.do_warning_exception_hangup_injection_here()
+
+            # get the provenance for this step:
+            prov = ds.get_provenance(self.pars.get_process_name(), self.pars.get_critical_pars(), session=session)
+
+            # try to find some measurements in memory or in the database:
+            measurements_list = ds.get_measurements(prov, session=session)
+
+            # note that if measurements_list is found, there will not be an all_measurements appended to datastore!
```

Comment:
Thinking ahead -- when we get to the CutoutsFile, all the stuff that would get stored in measurements (or the merged measurements/cutouts table) should be in the hdf5 cutouts file, so they are still accessible somewhere.

Reply:
I'm not sure I want to store the measurements along with the cutouts file in one process. That would mean that if you want to re-run the measuring part, you'd need to make a new cutouts HDF5 file. Seems a bit wasteful. Maybe it would be better to have a step for making cutouts and putting them into a file, and then making measurements from that file. The HDF5 files are like 10MB so they are not negligible.

Reply:
Hmm.  OK, here are my goals:

(1) The measurements table in the database only has the measurements that passed the preliminary cuts.  This is to keep that table manageable; at least with the stuff we saw in the stress test, doing that cuts down the length of that table by two orders of magnitude.  This is a table I expect us to be searching fairly frequently (any thing that does a search for n detections with xyz properties is going to search it), there's value in not letting this table bloat.

(2) Having the measurements available *somewhere*, in case we want to think about what happened with the preliminary cuts.  From a "at the end" point of view, having them in the cutouts file seems like the most obvious place.  But, you're right, that would mean redoing cutouts whenever redoing measurements.

So, more thought required.